README.md
Yii2 Field widgets
==============

## Introduction

[![Latest Stable Version](https://poser.pugx.org/itstructure/yii2-field-widgets/v/stable)](https://packagist.org/packages/itstructure/yii2-field-widgets)
[![Latest Unstable Version](https://poser.pugx.org/itstructure/yii2-field-widgets/v/unstable)](https://packagist.org/packages/itstructure/yii2-field-widgets)
[![License](https://poser.pugx.org/itstructure/yii2-field-widgets/license)](https://packagist.org/packages/itstructure/yii2-field-widgets)
[![Total Downloads](https://poser.pugx.org/itstructure/yii2-field-widgets/downloads)](https://packagist.org/packages/itstructure/yii2-field-widgets)
[![Build Status](https://scrutinizer-ci.com/g/itstructure/yii2-field-widgets/badges/build.png?b=master)](https://scrutinizer-ci.com/g/itstructure/yii2-field-widgets/build-status/master)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/itstructure/yii2-field-widgets/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/itstructure/yii2-field-widgets/?branch=master)

This is form field's widgets for the yii2 framework with the next field types:

- text
- textarea
- ckeditor
- ckeditorAdmin
- file
- checkbox
- dropdown
- password

## Dependencies

- php >= 7.1
- composer

## Installation

### Base install

Via composer:

`composer require "itstructure/yii2-field-widgets": "~1.2.4"`

or in section **require** of composer.json file set the following:

```json
"require": {
    "itstructure/yii2-field-widgets": "~1.2.4"
}
```

and command `composer install`, if you install yii2 project extensions first,

or command `composer update`, if all yii2 project extensions are already installed.

### If you are testing this package from local server directory

In application `composer.json` file set the repository, like in example:

```json
"repositories": [
    {
        "type": "path",
        "url": "../yii2-field-widgets",
        "options": {
            "symlink": true
        }
    }
],
```

Here,

**yii2-field-widgets** - directory name, which has the same directory level like application and contains yii2 field widgets package.

Then run command:

`composer require itstructure/yii2-field-widgets:dev-master --prefer-source`

## Usage

### Requirements

That widgets are designed to work in form with an active **model**, which is inherited from 
yii\db\ActiveRecord.

### Single mode

To use this mode, don't set `languageModel`. That will be **null**.

### Multilanguage mode

All fields will be with a language postfix:

**title_en**

**description_en**

**title_ru**

**description_ru**, e t. c.

For this mode it's necessary to have Language model with some of languages records.

Example:

```php
$form = ActiveForm::begin();
```

```php
echo Fields::widget([
    'fields' => [
        [
            'name' => 'title',
            'type' => FieldType::FIELD_TYPE_TEXT,
        ],
        [
            'name' => 'description',
            'type' => FieldType::FIELD_TYPE_CKEDITOR_ADMIN,
            'preset' => 'full',
            'options' => [
                'filebrowserBrowseUrl' => '/ckfinder/ckfinder.html',
                'filebrowserImageBrowseUrl' => '/ckfinder/ckfinder.html?type=Images',
                'filebrowserUploadUrl' => '/ckfinder/core/connector/php/connector.php?command=QuickUpload&type=Files',
                'filebrowserImageUploadUrl' => '/ckfinder/core/connector/php/connector.php?command=QuickUpload&type=Images',
                'filebrowserWindowWidth' => '1000',
                'filebrowserWindowHeight' => '700',
            ]
        ],
    ],
    'model'         => $model,
    'form'          => $form,
    'languageModel' => new Language()
]) ?>
```

```php
echo Html::submitButton($model->isNewRecord ? 'Create' : 'Update', ['class' => $model->isNewRecord ? 'btn btn-success' : 'btn btn-primary']) ?>
```

```php
ActiveForm::end();
```

## License

Copyright © 2018-2023 Andrey Girnik girnikandrey@gmail.com.

Licensed under the [MIT license](http://opensource.org/licenses/MIT). See LICENSE.txt for details.


changelog.md
### CHANGE LOG:

**1.2.4 January 29, 2023:**
- Upgrade copyright year.

**1.2.3 June 24, 2020:**
- Modify README syntax.

**1.2.2 August 10, 2018:**
- Change display translates in ```tableMultilanguage``` file by tabs.

**1.2.1 August 9, 2018:**
- Code fixes according with the PSR standards.

**1.2.0 May 12, 2018:**
- Modify dependencies: minimum-stability is set to dev.
- Set prefer-stable is true
- Add .scrutinizer file.
- Add badges:
    - Latest Stable Version
    - Latest Unstable Version
    - License
    - Total Downloads.
    - Build Status
    - Scrutinizer Code Quality

**1.1.0 February 13, 2018:**
- Depricate className() function. Use WidgetClass::class.

**1.0.0 February 4, 2018:**
- Create form field's widgets for the yii2 framework with the next field types:
    - text
    - textarea
    - ckeditor
    - ckeditorAdmin
    - file
    - checkbox
    - dropdown
    - password

- Create documentation.


src/Fields.php
<?php

namespace Itstructure\FieldWidgets;

use Itstructure\FieldWidgets\interfaces\LanguageListInterface;
use yii\base\{Widget, Model};
use yii\widgets\ActiveForm;

/**
 * Class Fields
 * Widget class for displaying a list of form fields with given types.
 *
 * @property array $fields
 * @property LanguageListInterface|null $languageModel
 * @property Model $model
 * @property ActiveForm $form
 *
 * @package Itstructure\FieldWidgets
 */
class Fields extends Widget
{
    /**
     * List of form fields.
     * You must set the following parameters for each field:: name, type.
     * name (string) - field name, example: title, description.
     * type (string) - field type, may be: text, textarea, ckeditor, 'ckeditorAdmin', file, checkbox, dropdown,
     * password.
     * Optional:
     * label (string) - field label.
     * data (array) - data to be inserted in to the field.
     * options (array) - field options.
     * hide (bool) - to switch off the field. Default false.
     * Example:
     * $fields = [
     *      [
     *          'name' => 'title',
     *          'type' => 'text'
     *      ],
     *      [
     *          'name' => 'description',
     *          'type' => 'ckeditor',
     *          'label' => 'Description',
     *          'preset' => 'custom',
     *          'options' => [
     *              'toolbarGroups' => [
     *                  [
     *                      'name' => 'undo'
     *                  ],
     *                  [
     *                      'name' => 'basicstyles',
     *                      'groups' => ['basicstyles', 'cleanup']
     *                  ],
     *                  [
     *                      'name' => 'colors'
     *                  ],
     *                  [
     *                      'name' => 'links',
     *                      'groups' => ['links', 'insert']
     *                  ],
     *                  [
     *                      'name' => 'others',
     *                      'groups' => ['others', 'about']
     *                  ],
     *                  ],
     *              'filebrowserBrowseUrl' => '/ckfinder/ckfinder.html',
     *              'filebrowserImageBrowseUrl' => '/ckfinder/ckfinder.html?type=Images',
     *              'filebrowserUploadUrl' => '/ckfinder/core/connector/php/connector.php?command=QuickUpload&type=Files',
     *              'filebrowserImageUploadUrl' => '/ckfinder/core/connector/php/connector.php?command=QuickUpload&type=Images',
     *              'filebrowserWindowWidth' => '1000',
     *              'filebrowserWindowHeight' => '700',
     *          ]
     *      ],
     *      [
     *          'name' => 'roles',
     *          'type' => 'checkbox',
     *          'label' => 'Roles',
     *          'data' => ArrayHelper::map($roles, 'name', 'name'),
     *          'options' => ['style' => 'width: 50%;'],
     *      ],
     * ];
     *
     * @var array
     */
    public $fields = [];

    /**
     * Language model.
     * For multilanguage mode.
     *
     * @var LanguageListInterface
     */
    private $languageModel = null;

    /**
     * Model with fields.
     *
     * @var Model
     */
    private $model;

    /**
     * Form object.
     *
     * @var ActiveForm
     */
    private $form;

    /**
     * Set the language model.
     * For multilanguage mode.
     *
     * @param mixed $languageModel
     *
     * @return void
     */
    public function setLanguageModel(LanguageListInterface $languageModel): void
    {
        $this->languageModel = $languageModel;
    }

    /**
     * Set model.
     *
     * @param Model $model
     *
     * @return void
     */
    public function setModel(Model $model): void
    {
        $this->model = $model;
    }

    /**
     * Set form.
     *
     * @param mixed $form
     *
     * @return void
     */
    public function setForm(ActiveForm $form): void
    {
        $this->form = $form;
    }

    /**
     * Starts the output widget of the form fields list with the specified types.
     *
     * @return string
     */
    public function run(): string
    {
        $params = [
            'fields' => $this->fields,
            'model'  => $this->model,
            'form'   => $this->form,
        ];

        if ($this->languageModel === null) {
            $template = 'fieldsSimple';
        } else {
            $template = 'fieldsMultilanguage';
            $params['languageModel'] = $this->languageModel;
        }

        return $this->render($template, $params);
    }
}
